bftengine/src/bcstatetransfer/RVBManager.cpp

```diff
@@ -770,45 +775,56 @@ void RVBManager::reportLastAgreedPrunableBlockId(uint64_t lastAgreedPrunableBloc
   }
 
   uint64_t current_rvb_id = start_rvb_id;
-  int32_t num_digests_added{0};
-  while (current_rvb_id <= lastAgreedPrunableBlockId) {
-    Digest digest;
-    if (!pruned_blocks_digests_.empty() && (current_rvb_id <= pruned_blocks_digests_.back().first)) {
+  // Avoid fetching already existing digests of to-be-pruned blocks
+  if (!pruned_blocks_digests_.empty()) {
+    while ((current_rvb_id <= lastAgreedPrunableBlockId) and (current_rvb_id <= pruned_blocks_digests_.back().first)) {
       current_rvb_id += config_.fetchRangeSize;
-      continue;
-    }
-    if (!as_->getPrevDigestFromBlock(current_rvb_id + 1,
-                                     reinterpret_cast<StateTransferDigest*>(digest.getForUpdate()))) {
-      LOG_FATAL(logger_,
-                "Digest not found:" << KVLOG(lastAgreedPrunableBlockId,
-                                             initial_size,
-                                             start_rvb_id,
-                                             current_rvb_id,
-                                             num_digests_added,
-                                             pruned_blocks_digests_.size()));
-      ConcordAssert(false);
     }
-    pruned_blocks_digests_.push_back(std::make_pair(current_rvb_id, std::move(digest)));
-    current_rvb_id += config_.fetchRangeSize;
-    ++num_digests_added;
   }
 
-  if (initial_size != pruned_blocks_digests_.size()) {
+  if (current_rvb_id < genesis_block_id) {
+    LOG_WARN(logger_, "Current pruning report has no impact on RVB data:" << KVLOG(current_rvb_id, genesis_block_id));
+    return;
+  }
+
+  const auto max_rvb_id = config_.fetchRangeSize * (lastAgreedPrunableBlockId / config_.fetchRangeSize);
+  auto num_rvb_blocks = (max_rvb_id - current_rvb_id) / config_.fetchRangeSize;
+  num_rvb_blocks += (current_rvb_id <= max_rvb_id) ? 1 : 0;
```

Comment:
I don't understand this calculation. I would expect that you add '1' to num_rvb_blocks in case you have a reminder in the computation in the line above.

Reply:
Thanks Yulia.

Remainder would not work in below scenario.
config_.fetchRangeSize = 24
lastAgreedPrunableBlockId = 825

Existing calculation is correct and it has went through extensive testing.